python/ray/tune/execution/ray_trial_executor.py

```diff
@@ -310,43 +316,79 @@ def _stage_and_update_status(self, trials: Iterable[Trial]):
 
         Stages placement groups of all trials.
         """
-        if not self._has_cleaned_up_pgs:
-            # Clean up existing placement groups after trigger the tuning
-            # run step() method for the first time
-            self._pg_manager.cleanup_existing_pg()
-            self._has_cleaned_up_pgs = True
-
         for trial in trials:
+            if len(self._staged_trials) >= self._max_staged_actors:
+                break
+
             if trial.status not in (Trial.PENDING, Trial.PAUSED):
                 continue
+
             if trial in self._staged_trials:
                 continue
-            if self._pg_manager.trial_in_use(trial):
-                continue
 
-            if not self._pg_manager.stage_trial_pg(trial):
-                # Break if we reached the limit of pending placement groups.
-                break
+            resource_request = trial.placement_group_factory
+
             self._staged_trials.add(trial)
+            self._resource_manager.request_resources(resource_request=resource_request)
 
-        self._pg_manager.update_status()
+        self._resource_manager.update_state()
 
-    def get_staged_trial(self):
-        """Get a trial whose placement group was successfully staged.
+    def get_ready_trial(self) -> Optional[Trial]:
+        """Get a trial whose resources are ready and that thus can be started.
 
         Can also return None if no trial is available.
 
         Returns:
             Trial object or None.
 
         """
-        # TODO(xwjiang): This method should consider `self._cached_actor_pg`.
         for trial in self._staged_trials:
-            if self._pg_manager.has_ready(trial):
+            resource_request = trial.placement_group_factory
+            # If we have a cached actor for these resources, return
+            if self._cached_resources_to_actor[resource_request]:
+                return trial
+
+            # If the resources are available from the resource manager, return
+            if self._resource_manager.has_resources_ready(
+                resource_request=resource_request
+            ):
                 return trial
 
         return None
 
+    def _maybe_use_cached_actor(self, trial, logger_creator) -> Optional:
```

Comment:
When a cached actor is reused, do we need to pop it from newly_added_xxx, in case it's there?

Reply:
As discussed offline, newly added actors are removed from the set in the same iteration they are added